File: src/300_最长递增子序列.rs
// 300_最长递增子序列
// https://leetcode.cn/problems/longest-increasing-subsequence/description/

impl Solution {
    pub fn length_of_lis(nums: Vec<i32>) -> i32 {
        let mut tails = vec![nums[0]];
    
        for &n in &nums[1..] {
            let mut l = 0;
            let mut r = tails.len(); // Attention
            while l < r {
                let mid = l + (r - l) / 2;
                if tails[mid] < n {
                    l = mid + 1;
                } else {
                    r = mid;
                }
            }
    
            if l == tails.len() {
                tails.push(n)
            } else {
                tails[l] = n
            }
        }   
    
        tails.len() as i32    
    }
}

File: src/300_最长递增子序列.cpp
#include <vector>
using namespace std;

// 300. 最长递增子序列
// https://leetcode.cn/problems/longest-increasing-subsequence/description/?envType=study-plan-v2&envId=top-interview-150
class Solution
{
public:
    int lengthOfLIS(vector<int> &nums)
    {
        int n = nums.size();
        if (n == 0)
        {
            return 0;
        }
        vector<int> dp(n, 1);
        int res = 1;
        for (int i = 1; i < n; i++)
        {
            for (int j = 0; j < i; j++)
            {
                if (nums[i] > nums[j])
                {
                    dp[i] = max(dp[i], dp[j] + 1);
                }
            }
            res = max(res, dp[i]);
        }
        return res;
    }
};